File: scripts/util/build_stamp.py
"""Content-hash freshness stamps for build steps — Python sibling of
build-stamp.ts, same stamp format. See scripts/util/README.md."""

from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
from typing import Iterable, Mapping, Optional

from scripts.util.paths import REPO_ROOT

STAMP_DIR = REPO_ROOT / "build" / "stamps"

InputHashes = Mapping[str, Optional[str]]

_HASH_CHUNK_BYTES = 1 << 22


def stamp_path(name: str) -> Path:
    return STAMP_DIR / f"{name}.json"


def hash_file(path: Path) -> str:
    h = hashlib.sha1()
    with path.open("rb") as f:
        while chunk := f.read(_HASH_CHUNK_BYTES):
            h.update(chunk)
    return h.hexdigest()


def input_hashes(paths: Iterable[Path]) -> dict[str, Optional[str]]:
    return {
        Path(os.path.relpath(p, REPO_ROOT)).as_posix(): hash_file(p) if p.exists() else None
        for p in sorted({p.resolve() for p in paths})
    }


def read_stamp(stamp: Path) -> Optional[InputHashes]:
    if not stamp.exists():
        return None
    return json.loads(stamp.read_text())["inputs"]


def stamp_is_current(stamp: Path, hashes: InputHashes, outputs: Iterable[Path]) -> bool:
    recorded = read_stamp(stamp)
    return recorded is not None and all(o.exists() for o in outputs) and dict(recorded) == dict(hashes)


def clear_stamp(stamp: Path) -> None:
    """Must run before a build writes any output, so a build that dies midway
    leaves no stamp vouching for its partial outputs."""
    stamp.unlink(missing_ok=True)


def write_stamp(stamp: Path, hashes: InputHashes) -> None:
    stamp.parent.mkdir(parents=True, exist_ok=True)
    stamp.write_text(json.dumps({"inputs": dict(hashes)}, indent=2, sort_keys=True) + "\n")

File: scripts/util/build-stamp.ts
// Content-hash freshness stamps for build steps. See scripts/util/README.md.

import { createHash } from 'node:crypto';
import {
  closeSync, existsSync, mkdirSync, openSync, readFileSync, readSync, rmSync, writeFileSync,
} from 'node:fs';
import { dirname, relative, resolve } from 'node:path';

import { REPO_ROOT } from './paths';

export const STAMP_DIR = resolve(REPO_ROOT, 'build/stamps');

/** Repo-relative path → sha1 of its content, or null when the file is absent. */
export type InputHashes = Readonly<Record<string, string | null>>;

export function stampPath(name: string): string {
  return resolve(STAMP_DIR, `${name}.json`);
}

const HASH_CHUNK_BYTES = 1 << 22;

export function hashFile(path: string): string {
  const hash = createHash('sha1');
  const buf = Buffer.allocUnsafe(HASH_CHUNK_BYTES);
  const fd = openSync(path, 'r');
  try {
    for (let n; (n = readSync(fd, buf)) > 0;) hash.update(buf.subarray(0, n));
  } finally {
    closeSync(fd);
  }
  return hash.digest('hex');
}

export function inputHashes(paths: readonly string[]): InputHashes {
  const hashes: Record<string, string | null> = {};
  for (const p of [...new Set(paths)].sort()) {
    hashes[relative(REPO_ROOT, p)] = existsSync(p) ? hashFile(p) : null;
  }
  return hashes;
}

export function readStamp(stamp: string): InputHashes | null {
  if (!existsSync(stamp)) return null;
  return (JSON.parse(readFileSync(stamp, 'utf8')) as { inputs: InputHashes }).inputs;
}

function sameHashes(a: InputHashes, b: InputHashes): boolean {
  const keys = Object.keys(a);
  return keys.length === Object.keys(b).length && keys.every((k) => a[k] === b[k]);
}

export function stampIsCurrent(
  stamp: string, hashes: InputHashes, outputs: readonly string[],
): boolean {
  const recorded = readStamp(stamp);
  return recorded !== null && outputs.every(existsSync) && sameHashes(recorded, hashes);
}

/** Must run before a build writes any output, so a build that dies midway
 *  leaves no stamp vouching for its partial outputs. */
export function clearStamp(stamp: string): void {
  rmSync(stamp, { force: true });
}

export function writeStamp(stamp: string, hashes: InputHashes): void {
  mkdirSync(dirname(stamp), { recursive: true });
  writeFileSync(stamp, `${JSON.stringify({ inputs: hashes }, null, 2)}\n`);
}
